xwiki-commons-core/xwiki-commons-observation/xwiki-commons-observation-local/src/main/java/org/xwiki/observation/internal/DefaultObservationManager.java

```diff
@@ -221,13 +221,15 @@ public void addListener(EventListener eventListener)
     @Override
     public void removeListener(String listenerName)
     {
-        synchronized (this.listenersByEvent) {
-            getListenersByName().remove(listenerName);
-            for (Map.Entry<Class<? extends Event>, Map<String, RegisteredListener>> entry : this.listenersByEvent
-                .entrySet()) {
-                entry.getValue().remove(listenerName);
-                if (entry.getValue().isEmpty()) {
-                    this.listenersByEvent.remove(entry.getKey());
+        if (this.listenersByEvent != null) {
```

Comment:
This if is wrong. It makes impossible to remove one of the registered listeners until some other manipulation of the ObservationManager is done (send an event, add another listener).

You just need to lock the return of getListenersByName() instead of this.listenersByEvent directly.

Reply:
Why not simply initialize the 2 maps in the default constructor or in the field definitions? Why do we need to rely on the lazy `initializeListeners` method instead of making sure we always have non-null maps already created?

Even better, instead of adding an "isInitialized" flag (to replace the null check that we are using now and is having side-effects), we could simply make the component `implement Initializable` and do both the map creations and the lookup of the listeners (i.e. calling `initializeListeners`) inside the `initialize` method.

WDYT?

Reply:
Re lazy init, see https://github.com/xwiki/xwiki-commons/blob/ac8c331e133a6480bbf08c92603a6c91bbd5eb8a/xwiki-commons-core/xwiki-commons-observation/xwiki-commons-observation-local/src/main/java/org/xwiki/observation/internal/DefaultObservationManager.java#L150

Reply:
Then maybe we need to use the `isInitialized` flag, as I mentioned, instead of the null check on the maps. This frees the maps to always be non-null and we can avoid the NPE as well.

Reply:
actually right now I'm trying to remove the lock on the map, and to lock the getter methods on the map instead. Taking some time because I want to test it properly :)